### PR TITLE
Distinguish between trusted and untrusted signatures and keys.

### PR DESCRIPTION
**Pull request by DemiMarie**
RPM needs to be able to differentiate between “key not found” (`RPMRC_NOKEY`) and “key or signature not trusted” (`RPMRC_NOTTRUSTED`).  A signature made by a revoked subkey is very much the latter situation.

### CODE REVIEW

**Comment by pmatilai** [comment]
Demi, development of the OpenPGP parser is frozen now, remember?

**Comment by pmatilai** [comment]
As a case, this may well be something rpm needs to handle, but going forward (ie assuming Sequoia and/or other "proper" implementation) I only see the in-tree parser maintained stripped to bare bone with no subkey support. So really, don't work on it anymore.

**Comment by DemiMarie** [comment]
> Demi, development of the OpenPGP parser is frozen now, remember?

Good point.  Does this also apply to the keyring code and similar?  My understanding is that more work is needed there, such as to using fingerprints instead of key IDs.  Also, I think the tests that this PR includes are valuable; they should pass for Sequoia-based RPM as well.

In any case, I could see this PR being useful for downstreams that can’t use Rust or need something they can backport.

**Comment by nwalfield** [comment]
In the very least, it would be good if the test case were merged.

**Comment by DemiMarie** [comment]
> In the very least, it would be good if the test case were merged.

The test case fails with current RPM, as RPM currently doesn’t know when it needs to distrust a signature.  This is still nowhere close to a full OpenPGP implementation (that’s orders of magnitude more complex), but it _is_ enough to make `gpg2 --export-options export-minimal --armor --output=output.asc -- '<TRUSTED_FINGERPRINT>' && rpmkeys --import output.asc` safe (at least ignoring expired subkeys, which are a trivial add-on) and that’s what I mostly care about.

**Comment by DemiMarie** [comment]
@pmatilai would you be okay with merging this change so that the test case can be included, and so that LTS distributions have something they can backport if they wish?

**Comment by nwalfield** [comment]
I'm a bit concerned about the new certificates added in 126d8962b.  The certificates will expire in two years.  If an OpenPGP backend handles expiration, then I think these tests will begin to fail.  It would probably be better to recreate the certificates without an expiration time.

```
$ sq inspect first-signing-key-revoked.asc
first-signing-key-revoked.asc: OpenPGP Certificate.

    Fingerprint: C9E4585A5D0201A9A077CE60D115BE9BEBDDD264
Public-key algo: EdDSA Edwards-curve Digital Signature Algorithm
Public-key size: 256 bits
  Creation time: 2022-03-26 22:10:10 UTC
Expiration time: 2024-03-25 22:10:10 UTC (creation time + P730D)
...
$ sq inspect rpm.org-ed25519-subkey-2-test.pub
rpm.org-ed25519-subkey-2-test.pub: OpenPGP Certificate.

    Fingerprint: 7B32CBC30D3423162732DD859DD6F039A0E0C1C7
Public-key algo: EdDSA Edwards-curve Digital Signature Algorithm
Public-key size: 256 bits
  Creation time: 2022-03-26 18:42:08 UTC
Expiration time: 2024-03-25 18:42:08 UTC (creation time + P730D)
...
$ sq inspect rpm.org-ed25519-subkey-test.pub
rpm.org-ed25519-subkey-test.pub: OpenPGP Certificate.

    Fingerprint: 7B32CBC30D3423162732DD859DD6F039A0E0C1C7
Public-key algo: EdDSA Edwards-curve Digital Signature Algorithm
Public-key size: 256 bits
  Creation time: 2022-03-26 18:42:08 UTC
Expiration time: 2024-03-25 18:42:08 UTC (creation time + P730D)
```

**Comment by DemiMarie** [comment]
> I'm a bit concerned about the new certificates added in [126d896](https://github.com/rpm-software-management/rpm/commit/126d8962b916fc81b187043a5d7922cfdc484d1c). The certificates will expire in two years. If an OpenPGP backend handles expiration, then I think these tests will begin to fail. It would probably be better to recreate the certificates without an expiration time.

Good idea.  Does the logic look good?

**Comment by DemiMarie** [comment]
Rebased on top of #2008 because otherwise #2004 breaks the tests.

**Comment by pmatilai** [comment]
NOTTRUSTED does seem like an entirely reasonable way to handle eg revoked subkeys as such, but sometimes sane behavior doesn't quite fit with the mad underlying OK/NOKEY etc semantics in rpm. I think this will require some more chewing before swallowing, if only because it quite fundamentally changes the landscape of rpm signature handling.

Up to now, imported == trusted since beginning of times, and there has never been a single case where rpm returns NOTTRUSTED. Which means that callers have no clear idea what it means and where they might encounter it, and so there almost certainly will be code in and out of rpm which will behave in unexpected/unwanted ways when presented with something else than OK/FAIL/NOKEY return. This also calls for test for the case where an installed packages has an NOTTRUSTED signature.

**Comment by nwalfield** [comment]
[Here, @DemiMarie wrote](https://github.com/rpm-software-management/rpm/issues/1978#issuecomment-1097712857):

> I suggest fixing these tests in the internal parser, provided that tests are added that make sure the Sequoia backend won’t actually verify a signature using the bad subkey. My PR https://github.com/rpm-software-management/rpm/pull/1993 should do most of it.

[I created three tests](https://github.com/nwalfield/rpm/tree/demi-handle-revocation):

  - A package signed with a subkey
  - A package signed with an expired subkey
  - A package signed with a revoked subkey

Using the Sequoia backend, all three of these tests pass.  It seems that not only is the internal OpenPGP parser unable to deal with this, but it can't even import the certificate:

```
269. rpmsigdig.at:304: testing rpmkeys --import, signed with a good subkey ...
../../tests/rpmsigdig.at:306:


if ! [ -d testing/ ]; then
    cp -aP "${RPMTEST}" .
    chmod -R u+w testing/
    mkdir -p testing/build
    ln -s ../data/SOURCES testing/build/
fi
export RPMTEST="${PWD}/testing"
export TOPDIR="${RPMTEST}/build"
export HOME="${RPMTEST}"

rm -rf "${RPMTEST}"`rpm --eval '%_dbpath'`/*
runroot rpm --initdb


echo Checking package before importing key:
runroot rpmkeys --define '_pkgverify_level all' -Kv /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
echo Importing key:
runroot rpmkeys --import /data/keys/alice.asc; echo $?
echo Checking for key:
runroot rpm -qi gpg-pubkey-eb04e625-* | grep Version
echo Checking package after importing key:
runroot rpmkeys --define '_pkgverify_level all' -Kv /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
echo Checking package after importing key, no digest:
runroot rpmkeys --define '_pkgverify_level all' -Kv --nodigest /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
echo Checking package after importing key, no signature:
runroot rpmkeys --define '_pkgverify_level all' -Kv --nosignature /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?

--- /dev/null	2022-03-31 08:23:21.524275212 +0000
+++ /home/us/neal/work/pep/rpm/b/tests/rpmtests.dir/at-groups/269/stderr	2022-04-13 20:27:07.767226325 +0000
@@ -0,0 +1 @@
+error: /data/keys/alice.asc: key 1 import failed.
--- -	2022-04-13 20:27:08.159403250 +0000
+++ /home/us/neal/work/pep/rpm/b/tests/rpmtests.dir/at-groups/269/stdout	2022-04-13 20:27:08.151227005 +0000
@@ -10,22 +10,23 @@
     MD5 digest: OK
 1
 Importing key:
-0
+1
...
```

Any idea what is going on here?  Is it perhaps because the internal parser rejects certificates, because the primary key is not signing capable?

**Comment by DemiMarie** [comment]
> [Here, @DemiMarie wrote](https://github.com/rpm-software-management/rpm/issues/1978#issuecomment-1097712857):
> 
> > I suggest fixing these tests in the internal parser, provided that tests are added that make sure the Sequoia backend won’t actually verify a signature using the bad subkey. My PR #1993 should do most of it.
> 
> [I created three tests](https://github.com/nwalfield/rpm/tree/demi-handle-revocation):
> 
> * A package signed with a subkey
> * A package signed with an expired subkey
> * A package signed with a revoked subkey
> 
> Using the Sequoia backend, all three of these tests pass. It seems that not only is the internal OpenPGP parser unable to deal with this, but it can't even import the certificate:
> 
> ```
> 269. rpmsigdig.at:304: testing rpmkeys --import, signed with a good subkey ...
> ../../tests/rpmsigdig.at:306:
> 
> 
> if ! [ -d testing/ ]; then
>     cp -aP "${RPMTEST}" .
>     chmod -R u+w testing/
>     mkdir -p testing/build
>     ln -s ../data/SOURCES testing/build/
> fi
> export RPMTEST="${PWD}/testing"
> export TOPDIR="${RPMTEST}/build"
> export HOME="${RPMTEST}"
> 
> rm -rf "${RPMTEST}"`rpm --eval '%_dbpath'`/*
> runroot rpm --initdb
> 
> 
> echo Checking package before importing key:
> runroot rpmkeys --define '_pkgverify_level all' -Kv /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
> echo Importing key:
> runroot rpmkeys --import /data/keys/alice.asc; echo $?
> echo Checking for key:
> runroot rpm -qi gpg-pubkey-eb04e625-* | grep Version
> echo Checking package after importing key:
> runroot rpmkeys --define '_pkgverify_level all' -Kv /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
> echo Checking package after importing key, no digest:
> runroot rpmkeys --define '_pkgverify_level all' -Kv --nodigest /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
> echo Checking package after importing key, no signature:
> runroot rpmkeys --define '_pkgverify_level all' -Kv --nosignature /data/RPMS/hello-2.0-1.x86_64-signed-with-subkey.rpm; echo $?
> 
> --- /dev/null	2022-03-31 08:23:21.524275212 +0000
> +++ /home/us/neal/work/pep/rpm/b/tests/rpmtests.dir/at-groups/269/stderr	2022-04-13 20:27:07.767226325 +0000
> @@ -0,0 +1 @@
> +error: /data/keys/alice.asc: key 1 import failed.
> --- -	2022-04-13 20:27:08.159403250 +0000
> +++ /home/us/neal/work/pep/rpm/b/tests/rpmtests.dir/at-groups/269/stdout	2022-04-13 20:27:08.151227005 +0000
> @@ -10,22 +10,23 @@
>      MD5 digest: OK
>  1
>  Importing key:
> -0
> +1
> ...
> ```
> 
> Any idea what is going on here? Is it perhaps because the internal parser rejects certificates, because the primary key is not signing capable?

Did you try with #1993?  Without #1993 RPM will reject a certificate that has a revoked subkey, if the revocation signature is the first one.  #1993 makes the behavior reasonable.

**Comment by DemiMarie** [comment]
@nwalfield you ran into a different bug: RPM is rejecting some of the signatures because they have critical subpackets it doesn’t understand.  I will file a patch to ignore the primary user ID flag.

**Comment by nwalfield** [comment]
I tried my test cases (09bce55a0) with your change (7e10e0d35), but I see that rpmkeys still can't import the certificate:

```
--- /dev/null	2022-03-31 08:23:21.524275212 +0000
+++ .../rpm/b/tests/rpmtests.dir/at-groups/269/stderr	2022-04-14 13:46:23.958169731 +0000
@@ -0,0 +1 @@
+error: /data/keys/alice.asc: key 1 import failed.
--- -	2022-04-14 13:46:24.093698347 +0000
+++ .../rpm/b/tests/rpmtests.dir/at-groups/269/stdout	2022-04-14 13:46:24.086169984 +0000
@@ -10,22 +10,23 @@
     MD5 digest: OK
 1
 Importing key:
-0
+1
...
```

**Comment by DemiMarie** [comment]
> I tried my test cases ([09bce55](https://github.com/rpm-software-management/rpm/commit/09bce55a013ed46908a5c4b36f88dee415080b17)) with your change ([7e10e0d](https://github.com/rpm-software-management/rpm/commit/7e10e0d3572b79a210b14c6b00da8e68af7d8c0f)), but I see that rpmkeys still can't import the certificate:
> 
> ```
> --- /dev/null	2022-03-31 08:23:21.524275212 +0000
> +++ .../rpm/b/tests/rpmtests.dir/at-groups/269/stderr	2022-04-14 13:46:23.958169731 +0000
> @@ -0,0 +1 @@
> +error: /data/keys/alice.asc: key 1 import failed.
> --- -	2022-04-14 13:46:24.093698347 +0000
> +++ .../rpm/b/tests/rpmtests.dir/at-groups/269/stdout	2022-04-14 13:46:24.086169984 +0000
> @@ -10,22 +10,23 @@
>      MD5 digest: OK
>  1
>  Importing key:
> -0
> +1
> ...
> ```

I suggest adding some debug logging in the subpacket parsing code to see which subpacket gets rejected.

**Comment by nwalfield** [comment]
> I suggest adding some debug logging in the subpacket parsing code to see which subpacket gets rejected.

[It's failing here](https://github.com/rpm-software-management/rpm/blob/7e10e0d3572b79a210b14c6b00da8e68af7d8c0f/rpmio/rpmpgp_internal.c#L329):

```
	    /* Reject duplicate key usage flags */
	    if (_digp->saved & PGPDIG_SIG_HAS_KEY_FLAGS)
		return 1;
```

HTH

**Comment by DemiMarie** [comment]
> > I suggest adding some debug logging in the subpacket parsing code to see which subpacket gets rejected.
> 
> [It's failing here](https://github.com/rpm-software-management/rpm/blob/7e10e0d3572b79a210b14c6b00da8e68af7d8c0f/rpmio/rpmpgp_internal.c#L329):
> 
> ```
> 	    /* Reject duplicate key usage flags */
> 	    if (_digp->saved & PGPDIG_SIG_HAS_KEY_FLAGS)
> 		return 1;
> ```
> 
> HTH

If the signature actually has more than one key usage subpacket in the hashed section, this is expected behavior, but I would also consider that to be a very strange signature.  Otherwise, something has gone wrong somewhere.  Can you check if the flag gets reset properly between signatures?

**Comment by nwalfield** [comment]
Using this MR plus https://github.com/rpm-software-management/rpm/pull/2026, the test keys that I generated can now be imported.

File: rpmio/rpmpgp_internal.c [hunk 1]
```diff
@@ -1179,7 +1179,11 @@ rpmRC pgpVerifySignature(pgpDigParams key, pgpDigParams sig, DIGEST_CTX hashctx)
 	pgpDigAlg ka = key->alg;
 	if (sa && sa->verify) {
 	    if (sa->verify(ka, sa, hash, hashlen, sig->hash_algo) == 0) {
-		res = RPMRC_OK;
+		if ((key->saved & PGPDIG_UNTRUSTED) ||
+		    (sig->saved & PGPDIG_UNTRUSTED))
+		    res = RPMRC_NOTTRUSTED;
+		else
+		    res = RPMRC_OK;
```
Comment (nwalfield) [inline rpmio/rpmpgp_internal.c:1204]: I agree with @pmatilai that we [shouldn't change the API](https://github.com/rpm-software-management/rpm/pull/1993#issuecomment-1097830727).

I think we can emit this type of information via [my proposed lint interface](https://github.com/rpm-software-management/rpm/issues/1974#issuecomment-1081779174).  We should, of course, consider returning richer error messages [when we transition to a new API](https://github.com/rpm-software-management/rpm/issues/1974#issuecomment-1081779174).

File: rpmio/rpmpgp_internal.c [hunk 2]
```diff
@@ -335,20 +331,42 @@ static int pgpPrtSubType(const uint8_t *h, size_t hlen, pgpSigType sigtype,
...
 	    break;
-	case PGPSUBTYPE_EXPORTABLE_CERT:
-	case PGPSUBTYPE_TRUST_SIG:
-	case PGPSUBTYPE_REGEX:
+	case PGPSUBTYPE_SIGNER_USERID: /* RFC4880 §5.2.3.22 Signer's User ID */
+	    impl = *p;
+	    break;
+	case PGPSUBTYPE_EMBEDDED_SIG: /* embedded signature */
+	    impl = *p;
+	    break;
```
Comment (nwalfield) [inline rpmio/rpmpgp_internal.c:339]: This code makes me suspect that rpm is not checking the [primary key binding signature](https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.1), which signing-capable subkeys must have to be valid.  Is that right?

File: rpmio/rpmpgp_internal.c [hunk 3]
```diff
@@ -335,20 +331,42 @@ static int pgpPrtSubType(const uint8_t *h, size_t hlen, pgpSigType sigtype,
...
+	case PGPSUBTYPE_REVOKE_REASON: /* RFC4880 §5.2.3.23 Reason for Revocation */
+	    if (plen < 2)
+		return 1; /* missing reason code */
+	    if (sigtype == PGPSIGTYPE_SUBKEY_REVOKE)
+		impl = *p;
```
Comment (nwalfield) [inline rpmio/rpmpgp_internal.c:344]: Why is the reason for revocation subpacket only understood in the context of subkey revocations and not certificate revocations or user id revocations?

File: rpmio/rpmpgp_internal.c [hunk 4]
```diff
@@ -999,12 +1006,8 @@ int pgpPrtParams(const uint8_t * pkts, size_t pktlen, unsigned int pkttype,
...
 
 	if (selfsig) {
-	    /* subkeys must be followed by binding signature */
-	    int xx = 1; /* assume failure */
-
-	    if (!(prevtag == PGPTAG_PUBLIC_SUBKEY &&
-		  selfsig->sigtype != PGPSIGTYPE_SUBKEY_BINDING))
-		xx = pgpVerifySelf(digp, selfsig, all, i);
+	    /* subkeys must be followed by binding or revocation signature */
+	    int xx = pgpVerifySelf(digp, selfsig, all, i, prevtag);
```
Comment (nwalfield) [inline rpmio/rpmpgp_internal.c:1040]: If we are willing to admit multiple self signatures, which I think the last commit and this one now allow, then we shouldn't fail if a self-sig can't be verified.  We should only fail if no self signatures can be verified.

**Review by nwalfield — COMMENTED** [review]
Overall, I think these changes are an improvement to the internal parser, but the complexity of the internal parser increasingly scares me.

**Review by DemiMarie — COMMENTED** [review]
The complexity of the internal parser scares me too, but I suspect most of that complexity is essential.  I wonder how GRUB’s parser fares.

**Comment by nwalfield** [comment]
I'll rework the patch series.

Is 6bf8f0350dedb8be5bcab94fdded73c00a37652d necessary? Also, this change isn't described by the commit message: https://github.com/DemiMarie/rpm/commit/6bf8f0350dedb8be5bcab94fdded73c00a37652d#diff-1b8795f690eee3ed483f0ce64cd81e4d42c681a13c98314a3250c4fac5159485R620

**Comment by DemiMarie** [comment]
> I'll rework the patch series.

Thank you.

> Is [6bf8f03](https://github.com/rpm-software-management/rpm/commit/6bf8f0350dedb8be5bcab94fdded73c00a37652d) necessary? Also, this change isn't described by the commit message: [DemiMarie@6bf8f03#diff-1b8795f690eee3ed483f0ce64cd81e4d42c681a13c98314a3250c4fac5159485R620](https://github.com/DemiMarie/rpm/commit/6bf8f0350dedb8be5bcab94fdded73c00a37652d#diff-1b8795f690eee3ed483f0ce64cd81e4d42c681a13c98314a3250c4fac5159485R620)

With the current code I believe it is a no-op, but it makes the code less error-prone and is an overall improvement.

File: rpmio/rpmpgp_internal.c (hunk 1 above)
Comment (nwalfield) [reply rpmio/rpmpgp_internal.c:1204]: If we return NOKEY, then I think the whole certificate is ignored, which is not what we want.

Comment (nwalfield) [reply rpmio/rpmpgp_internal.c:1204]: I think the better approach is to just ignore the subkey if it appears to be revoked, i.e., drop this change.

**Comment by pmatilai** [comment]
After way too much procrastinating and indecision...

We wont be able to axe the internal parser in 4.19, dunno about 4.20. But that's a long way off, and until then we need to support what we ship. To avoid having to pile this complexity on and on and on to deal with subkey this and subkey that, I think the only possible solution is to saw off the limb to save the patient: remove the subkey support (#2278). That's where like 90% of the complexities seem to be.

Clearly quite a bit of work has gone into this PR and I hate to throw that away, but it's not like I didn't warn right from the beginning. As most of this PR is dealing with subkey matters, merging it just to remove the added complexity right after doesn't make any sense. Lest it all go to waste: at least the bit to support Sequoia-generated keys deserve salvaging but that's really unrelated and should be filed separately. Ditto for test-cases, but only for the Sequoia backend. There may be others but lets first get the subkeys out, it'll be clearer then.

Thanks and apologies, I should've been firm with this right from the start. I'd like to claim the crystal ball on this all was still fuzzy in the spring but that'd be lame.

**Pull request closed without merging**